freqtrade/rpc/api_server.py

```diff
@@ -90,11 +87,43 @@ def send_msg(self, msg: str) -> None:
         # TODO: implement me
         raise NotImplementedError
 
+    def shutdown_api_server(self):
+        """
+        Stop the running flask application
+
+        Records the shutdown in logger.info
+        :return:
+        """
+        func = request.environ.get('werkzeug.server.shutdown')
+        if func is None:
+            raise RuntimeError('Not running the Flask Werkzeug Server')
+        if func is not None:
+            logger.info('Stopping the Local Rest Server')
+            func()
+            return
+
+    def cleanup(self) -> None:
+        """
+        Stops the running application server
+
+        Does not stop the thread,this may not be the desired outcome of cleanup. TBC
+        :return:
+        """
+        self.shutdown_api_server()
```

Comment:
I think we can't access `request.environ.get('werkzeug.server.shutdown')` because we are not in request scope, you can test this if you disable the api_server in config.json while it runs and then trigger `/reload_conf`. I'm also not entirely sure how to handle this. The best solution would be to start the app non-blocking and wait in a while loop, until some flag is changed via `cleanup()`, e.g.:

```
def run(self):
    # TODO: start non-blocking server
    while self.running:
        time.sleep(1)

def cleanup(self):
    self.running = False
```

Reply:
werkzeug.server.shutdow works for me, have tested lots of times calling 
http://127.0.0.1:5002/stop_api

I might be missing the detail in your comment.

Reply:
Yea if you stop it via the HTTP endpoint it works fine, but `cleanup` gets called when the bot crashes, someone reloads the config, etc.. And in this case you can't access `request` because we're not in a HTTP request context, therefore an exception will be raised.

Reply:
OK, thanks

I'll remove the call to stop api server run from the cleanup